### ISSUE TITLE
Add option to func-name-matching to allow names to include namespace prefixes

### ISSUE THREAD

**Issue by edg2s** [eslint/eslint#11360]
**What rule do you want to change?**
func-name-matching
**Does this change cause the rule to produce more or fewer warnings?**
could be more or fewer, depending on use case
**How will the change be implemented? (New option, new default behavior, etc.)?**
new option
**Please provide some example code that this change will affect:**
These names should be allowed with the rule enabled:

```js
App1.Widget = function App1Widget () { ... }

App2.Widget = function App2Widget () { ... }

Company1.App.Widget = function Company1AppWidget () { ... }

Company2.App.Widget = function Company2AppWidget () { ... }
```

**What does the rule currently do for this code?**
Currently the rule only matches the property name, so both functions would have to have the name 'Widget'. One of the reasons to name functions is to make stack traces more readable, and if different methods are forced to have the same name, that makes that feature less useful.

**What will the rule do after it's changed?**
Other naming schemes would be allowed, such as concatenating the namespace, or allowing for the variable name to only partially match the function name.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by edg2s** [comment]
FYI this is the naming scheme we use for Wikipedia's VisualEditor, and a few other pieces of software at Wikipedia: https://github.com/wikimedia/VisualEditor/blob/master/src/dm/ve.dm.Surface.js

It's quite common for developers to copy/paste a class and forget to rename the function, so this would be a useful rule for us.
